crates/qpy/src/value.rs

```diff
@@ -0,0 +1,604 @@
+
+pub mod bit_types {
+    pub const QUBIT: u8 = b'q';
+    pub const CLBIT: u8 = b'c';
+}
+
+#[derive(Debug)]
+pub struct QPYWriteData<'a> {
+    pub circuit_data: &'a mut CircuitData,
+    pub version: u32,
+    pub _use_symengine: bool, // TODO: remove this field?
+    pub clbits: &'a ObjectRegistry<Clbit, ShareableClbit>,
+    pub standalone_var_indices: HashMap<u128, u16>, // mapping from the variable's UUID to its index in the standalone variables list
+    pub annotation_handler: AnnotationHandler<'a>,
+}
+#[derive(Debug)]
+pub struct QPYReadData<'a> {
+    pub circuit_data: &'a mut CircuitData,
+    pub version: u32,
+    pub use_symengine: bool,
+    pub standalone_vars: HashMap<u16, qiskit_circuit::Var>,
+    pub standalone_stretches: HashMap<u16, qiskit_circuit::Stretch>,
+    pub vectors: HashMap<Uuid, (Py<PyAny>, Vec<u32>)>,
+    pub annotation_handler: AnnotationHandler<'a>,
+}
+
+// this is how tags for various value types are encoded in a QPY file
+pub mod tags {
+    pub const BOOL: u8 = b'b';
+    pub const INTEGER: u8 = b'i';
+    pub const FLOAT: u8 = b'f';
+    pub const COMPLEX: u8 = b'c';
+    pub const CASE_DEFAULT: u8 = b'd';
+    pub const REGISTER: u8 = b'R';
+    pub const RANGE: u8 = b'r';
+    pub const TUPLE: u8 = b't';
+    pub const NUMPY_OBJ: u8 = b'n';
+    pub const PARAMETER: u8 = b'p';
+    pub const PARAMETER_VECTOR: u8 = b'v';
+    pub const PARAMETER_EXPRESSION: u8 = b'e';
+    pub const STRING: u8 = b's';
+    pub const NULL: u8 = b'z';
+    pub const EXPRESSION: u8 = b'x';
+    pub const MODIFIER: u8 = b'm';
+    pub const CIRCUIT: u8 = b'q';
+}
+
+pub fn type_name(type_key: u8) -> String {
+    String::from(match type_key {
+        tags::BOOL => "boolean",
+        tags::INTEGER => "integer",
+        tags::FLOAT => "float",
+        tags::COMPLEX => "complex",
+        tags::CASE_DEFAULT => "case default",
+        tags::REGISTER => "register",
+        tags::RANGE => "range",
+        tags::TUPLE => "tuple",
+        tags::NUMPY_OBJ => "numpy object",
+        tags::PARAMETER => "parameter",
+        tags::PARAMETER_VECTOR => "parameter vector",
+        tags::PARAMETER_EXPRESSION => "parameter expression",
+        tags::STRING => "string",
+        tags::NULL => "null",
+        tags::EXPRESSION => "expression",
+        tags::MODIFIER => "modifier",
+        tags::CIRCUIT => "circuit",
+        _ => "unknown type",
+    })
+}
+
+pub mod modifier_types {
+    pub const INVERSE: u8 = b'i';
+    pub const CONTROL: u8 = b'c';
+    pub const POWER: u8 = b'p';
+}
+
+#[binrw]
+#[derive(Debug)]
+pub enum ExpressionType {
+    #[brw(magic = b'b')]
+    Bool,
+    #[brw(magic = b'u')]
+    Uint(u32),
+    #[brw(magic = b'f')]
+    Float,
+    #[brw(magic = b'd')]
+    Duration,
+}
+
+pub mod expression_var_declaration {
+    pub const INPUT: u8 = b'I';
+    pub const CAPTURE: u8 = b'C';
+    pub const LOCAL: u8 = b'L';
+    pub const STRETCH_CAPTURE: u8 = b'A';
+    pub const STRETCH_LOCAL: u8 = b'O';
+}
```

Comment:
Any reason why this is not an enum, like: 

```suggestion
#[binrw]
#[brw(repr = u8)]
#[repr(u8)]
#[derive(Debug)]
pub enum expression_var_declaration {
    INPUT = b'I',
    CAPTURE = b'C',
    LOCAL = b'L',
    STRETCH_CAPTURE = b'A',
    STRETCH_LOCAL = b'O',
}
```
And then in `formats.rs`:
```rust
pub struct ExpressionVarDeclarationPack {
    pub uuid_bytes: [u8; 16],
    pub usage: expression_var_declaration,
...
```

Reply:
I think I used to have a concrete reason for some of those, but I don't remember it; I'll try making the switch and see whether some problem pops up.

Reply:
Yeah, it worked. Maybe I didn't think of using the `#[repr(u8)]` tag at the time?